lib/tree.rb

```diff
@@ -16,47 +16,121 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(log(n)) where n is the number of nodes
+  # Space Complexity: O(1)
   def add(key, value)
-    raise NotImplementedError
+    new_node = TreeNode.new(key, value)
+    if !@root 
+      @root = new_node
+      return 
+    end
+    add_node(new_node, @root)
+
+  end
+  
+  def add_node(new_node, curr)
+    if curr.key >= new_node.key
+      if !curr.left 
+        curr.left = new_node
+        return
+      end
+      add_node(new_node, curr.left)
+    else
+      if !curr.right
+        curr.right = new_node
+        return
+      end
+      add_node(new_node, curr.right)
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+
+  # Time Complexity: O(log(n)) where n is the number of nodes
+  # Space Complexity: O(1)
   def find(key)
-    raise NotImplementedError
+    return if !@root
+    curr = @root
+    while curr 
+      if curr.key == key
+        return curr.value
+      elsif curr.key > key
+        curr = curr.left
+      else
+        curr = curr.right
+      end
+    end
+    return false
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n) where n is the number of nodes
+  # Space Complexity: O(n) ""
   def inorder
-    raise NotImplementedError
+    inorder_traversal(@root, [])
+  end
+
+  def inorder_traversal(node, node_list)
+    return node_list if !node
+    inorder_traversal(node.left, node_list)
+    visit(node, node_list)
+    inorder_traversal(node.right, node_list)
+  end
+
+  def visit(node, arr)
+    arr.push({key: node.key, value: node.value})
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n) where n is the number of nodes
+  # Space Complexity: O(n) ""
   def preorder
-    raise NotImplementedError
+    preorder_traversal(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def preorder_traversal(node, node_list)
+    return node_list if !node
+    visit(node, node_list)
+    preorder_traversal(node.left, node_list)
+    preorder_traversal(node.right, node_list)
+  end
+
+  # Time Complexity: O(n) where n is the number of nodes
+  # Space Complexity: O(n) "" 
   def postorder
-    raise NotImplementedError
+    postorder_traversal(@root, [])
+  end
+
+  def postorder_traversal(node, node_list)
+    return node_list if !node
+    postorder_traversal(node.left, node_list)
+    postorder_traversal(node.right, node_list)
+    visit(node, node_list)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n) where n is the number of nodes
+  # Space Complexity: O(n)
```

Comment:
The space complexity will be O(n) if the tree is unbalanced (due to the system stack) or O(log n) if the tree is balanced.